### ISSUE TITLE
Bug: [no-shadow] Doesn't catch shadowing of global types in generics

### ISSUE THREAD

**Issue by CreativeTechGuy** [typescript-eslint/typescript-eslint#5935]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.8.4&sourceType=module&code=MYGwhgzhAECyCeBhcUA8AJAKrAMgURAFMBbQgOwBcA+aAbwFgAoJgXyaYvgAdDpEBXCBQD2xTN14BeaGX7EARoQBOAbnaNQkGAmRbMAd2GoBQ0eJ40GzRm2uce0LLgIlyFaNKFKAlmQDmKkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uQAsBDACYcA7uiiJo0DtEjgwAXxDygA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

*See Playground link above*

### ESLint Config

*See Playground link above*

### tsconfig

*See Playground link above*

### Expected Result

I expected both the `HTMLElement` generic and the `HTMLElement` type declaration to both be an error as they both shadow the global type of the same name. The type declaration does show an error (actually a TypeScript compiler error) but there is no error for shadowing the type via a generic. I'd expect `no-shadow` to catch shadowing with generics.

### Actual Result

Only types which are defined in the module scope are identified as shadowing when used in generic declarations.

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `5.42.1` |
| `@typescript-eslint/parser`            | `5.42.1` |
| `TypeScript`                           | `4.8.4` |
| `ESLint`                               | `8.15.0` |

**Comment by bradzacher** [comment]
For comparison - the following similar value-space code is correctly reported:

```ts
{
  const foo = 1;
}
function foo() {}
```

I believe the rule needs to be taught that the ordering doesn't matter for types.
